### PR TITLE
Optical flow SAD code ported from PX4/Flow

### PR DESCRIPTION
**Pull request by jgoppert**
This ports the sum of difference optical flow code that is actually running on the PX4/Flow to use in the gazebo simulation for optical flow.

### CODE REVIEW

**Comment by jgoppert** [comment]
This closes #35 . I am flying with an increased flow camera rate of 100 hz, but with actual field of view and 64x64 pixels, working much better than before.

**Comment by mhkabir** [comment]
Nice! 

I'm curious as to whether this would work better in the real world than KLT with (narrow lens) bottom facing machine vision cameras...

**Comment by jgoppert** [comment]
I believe this is ready to merge, @bkueng can you please review?

File: include/flow.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+ *
+ ****************************************************************************/
+
+#ifndef FLOW_H_
+#define FLOW_H_
+
+#include <stdint.h>
+
+/**
+ * @brief Computes pixel flow from image1 to image2
```
Comment (bkueng) [inline include/flow.h:40]: The API doc could be extended to include an argument description. What's the image format, array sizes, output, value range etc. It's somewhat more described in the source file, but I think it should go here.

File: src/gazebo_opticalFlow_plugin.cpp [hunk 2]
```diff
@@ -23,13 +23,16 @@
...
 #include "gazebo_opticalFlow_plugin.h"
 
-#include <cv.h>
 #include <highgui.h>
 #include <math.h>
 #include <string>
 #include <iostream>
 #include <boost/algorithm/string.hpp>
 
+extern "C" {
```
Comment (bkueng) [inline src/gazebo_opticalFlow_plugin.cpp]: extern C needs to go into the headers

File: src/gazebo_opticalFlow_plugin.cpp [hunk 3]
```diff
@@ -146,62 +151,43 @@ void OpticalFlowPlugin::OnNewFrame(const unsigned char * _image,
...
-  if (!old_gray.empty() && featuresPrevious.size() > 0){
-    calcOpticalFlowPyrLK(old_gray, frame_gray, featuresPrevious, featuresNextPos, featuresFound, err, cv::Size(8,8), 0);
+  float x_gyro_rate = 0;
+  float y_gyro_rate = 0;
+  float z_gyro_rate = 0;
+  float x_flow = 0;
+  float y_flow = 0;
+
+  // check for correct image size
+  if (frame_gray.rows != 64 || frame_gray.cols != 64) {
```
Comment (bkueng) [inline src/gazebo_opticalFlow_plugin.cpp]: This check should go into the Load() method, as the size is not changed after startup.

**Review by bkueng — COMMENTED** [review]
- Can you rename settings.{h,c} to make it clear they belong to the flow plugin?

Looks good overall. One thing I was wondering: how do you select the gazebo world if you don't use ROS? It seems to be fixed to the model in the sitl run script.

**Comment by jgoppert** [comment]
@bkueng Thanks for the review, I think I fixed everything, going to trigger another build on my Firmware PR to check it works.

File: include/gazebo_opticalFlow_plugin.h [hunk 4]
```diff
@@ -34,6 +34,11 @@
 #include <iostream>
 #include <boost/timer/timer.hpp>
 
+extern "C" {
```
Comment (bkueng) [inline include/gazebo_opticalFlow_plugin.h:37]: @jgoppert there seems to have been a misunderstanding. what I meant was to move `extern "C"` into the `flow_settings.h` and `flow.h` headers around all function declarations. That's the common way to do it, as the user of the header should not need to bother if it's C or C++.

**Pull request merged**